app/models/webhook.rb

```diff
@@ -81,19 +90,50 @@ def ca_certs_store
 
   private
 
+  def http_headers_json
+    return if http_headers.blank?
+
+    JSON.parse(http_headers)
+  rescue JSON::ParserError => e
+    errors[:http_headers] << "is not valid JSON: #{e}"
+  end
+
   def set_default_template
     self.webhook_template = WebhookTemplate.find_by!(name: DEFAULT_PAYLOAD_TEMPLATE)
   end
 
-  def rendered_payload(event_name, payload)
-    webhook_template.render(
-      variables: {
-        event_name: event_name,
-        object: payload[:object],
-        context: payload[:context],
-        payload: payload,
-        webhook_id: id
-      }
+  def variables(event_name, payload)
+    {
+      event_name: event_name,
+      object: payload[:object],
+      context: payload[:context],
+      payload: payload,
+      webhook_id: id
+    }
+  end
+
+  def render_source(source, event_name, payload)
+    scope = Foreman::Renderer.get_scope(
+      klass: ForemanWebhooks::Renderer::Scope::WebhookTemplate,
+      variables: variables(event_name, payload),
+      source: source
     )
+    Foreman::Renderer.render(source, scope)
+  end
+
+  def rendered_payload(event_name, payload)
+    webhook_template.render(variables: variables(event_name, payload))
+  end
+
+  def rendered_headers(event_name, payload)
+    return nil if http_headers.empty?
+
+    source = Foreman::Renderer::Source::String.new(name: 'HTTP header template', content: http_headers)
+    render_source(source, event_name, payload)
+  end
+
+  def rendered_targed_url(event_name, payload)
+    source = Foreman::Renderer::Source::String.new(name: 'HTTP header template', content: target_url)
```

Comment:
Is it relevant that the name is the same as for headers?

Reply:
Overlook, fixed.

I also removed the order in the controller above to prevent conflicts.